lib/command.js

```diff
@@ -1181,20 +1242,27 @@ Expecting one of '${allowedValues.join("', '")}'`);
   /**
    * Once we have a promise we chain, but call synchronously until then.
    *
-   * @param {Promise|undefined} promise
+   * @param {*} chain
    * @param {Function} fn
-   * @return {Promise|undefined}
+   * @return {*}
    * @api private
    */
 
-  _chainOrCall(promise, fn) {
-    // thenable
-    if (promise && promise.then && typeof promise.then === 'function') {
+  _chainOrCall(chain, fn) {
+    if (isThenable(chain)) {
       // already have a promise, chain callback
-      return promise.then(() => fn());
+      return chain.then(fn);
     }
+
     // callback might return a promise
-    return fn();
+    chain = fn(chain);
+    if (isThenable(chain) && !this._asyncParsing) {
+      chain.then(x => x, () => {}); // prevent unhandled rejection
+      process.exitCode = 1; // additionally indicates to the user that user code (parsers, hooks, action) is not the error source
+      throw new Error(`.parse() is incompatible with async argParsers, hooks and actions.
```

Comment:
An alternative to throwing would be to log a warning and carry on, so preserving existing behaviour but encouraging more careful coding by author. This does increase risk authors miss the change in behaviour as unit tests still likely to pass, but does not hard-break end users. Thoughts?

(On that note, some authors have asked why they need to call `.parseAsync()` when `.parse()` seems to work fine for them.)

Reply:
I like the idea, see [385fe48](https://github.com/tj/commander.js/pull/1915/commits/385fe48b00ce0bfdab8702bf83e899893ef65b5f).